### PR TITLE
CI: Add check for SPDX license headers

### LINKED ISSUE #240: CI: add license checker
As noted by @grahamwhaley https://github.com/kata-containers/runtime/pull/196#discussion_r182019501, we should mandate that all new and modified files contain the correct [SPDX license identifer](https://spdx.org/licenses/Apache-2.0.html).

### PR DESCRIPTION
**Pull request by jodh-intel**
Add a check to ensure that all files where possible contain the
required SPDX license header.

Fixes #240.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

File: .ci/static-checks.sh [hunk 1]
```diff
@@ -175,6 +175,52 @@ check_versions()
...
+		--exclude="*.json" \
+		--exclude=".git/*" \
+		--exclude=".gitignore" \
+		--exclude="Gopkg.lock" \
+		--exclude="LICENSE" \
+		--exclude="protocols/grpc/*.pb.go" \
+		--exclude="vendor/*" \
+		--exclude="VERSION" \
+		--exclude="*.md" \
+		-EL "\<${pattern}\>" \
```
Comment (jodh-intel) [inline .ci/static-checks.sh]: This list of excludes works reasonably well, but I wonder if we want to add `.pullapprove.yml` to it too? We could of course instead add the SPDX header to that file ;)

File: .ci/static-checks.sh [hunk 2]
```diff
@@ -175,6 +175,52 @@ check_versions()
...
+	local -r spdx_license="Apache-2.0"
+	local -r pattern="${spdx_tag}: ${spdx_license}"
+
+	echo "INFO: Checking for SPDX license headers"
+
+	# List of files to check for a license header inside
+	files=$(git diff-tree \
+		--name-only \
+		--no-commit-id \
+		--diff-filter="ACMRUX" \
```
Comment (jcvenegas) [inline .ci/static-checks.sh]: ACMRUX ?

File: .ci/static-checks.sh (hunk 1 above)
Comment (jcvenegas) [reply .ci/static-checks.sh]: what about exclude yml and toml files in general?

Comment (jodh-intel) [reply .ci/static-checks.sh]: I guess we could, although for example the versions database does have the SPDX tag:

- https://github.com/kata-containers/runtime/blob/master/versions.yaml#L4

But yes, we'd need to add it to `configuration.toml` (and probably should ;)

File: .ci/static-checks.sh (hunk 2 above)
Comment (grahamwhaley) [reply .ci/static-checks.sh]: new to me - went to look it up - still digesting... https://git-scm.com/docs/git-diff-tree#git-diff-tree---diff-filterACDMRTUXB82308203

File: .ci/static-checks.sh (hunk 1 above)
Comment (grahamwhaley) [reply .ci/static-checks.sh]: I vote maybe we should add them if that does not upset any of the relevant processing tools or bots

File: .ci/static-checks.sh (hunk 2 above)
Comment (jodh-intel) [reply .ci/static-checks.sh]: Yeah, it's a bit opaque :) I've updated the file to make it a bit clearer ;)

**Comment by grahamwhaley** [comment]
and thanks @jodh-intel for writing that :-)

File: .ci/static-checks.sh (hunk 1 above)
Comment (jodh-intel) [reply .ci/static-checks.sh]: Ok - added.

**Comment by GabyCT** [comment]
lgtm

**Comment by jcvenegas** [comment]
@GabyCT @chavafg "jenkins-ci-ubuntu-16-04 Expected — Waiting for status to be reported" I see ubuntu did not reported, is this a known issue?

**Comment by chavafg** [comment]
Don't know what happened. The job finished successfully: http://kata-jenkins-ci.westus2.cloudapp.azure.com/job/kata-containers-tests-ubuntu-16-04-PR/325/
Restarted the job.

**Comment by grahamwhaley** [comment]
afaict, the CI did pass - at least as well as any other PR. @chavafg , looking at a few of the logs for the 16.04 builds, they seem to have a number of error-ish type lines (around yaml not found and coverage not reported etc.). I wonder if we can make the scripts a little smarter to reduce some of the noise in the future?
merging..

**Comment by bergwolf** [comment]
It seems that this PR is breaking all CIs. I'm seeing 
```
ERROR: Required license identifier ('SPDX-License-Identifier: Apache-2.0') missing from following files:
```
for PRs that do not add any new files. Do we want to block all PRs until the switch is done?

**Comment by grahamwhaley** [comment]
@bergwolf  (or anybody) - do you have a link to a PR that has the failing CI so we can have a look?

**Comment by bergwolf** [comment]
@grahamwhaley https://travis-ci.org/kata-containers/runtime/builds/368056793

**Comment by grahamwhaley** [comment]
Ah, I'm afraid the tool did exactly what it was meant to. The bit you did not paste is:
```
virtcontainers/agent.go
virtcontainers/api.go
virtcontainers/api_test.go
virtcontainers/hyperstart_agent.go
virtcontainers/implementation.go
virtcontainers/interfaces.go
virtcontainers/kata_agent.go
virtcontainers/noop_agent.go
virtcontainers/pkg/vcmock/mock.go
virtcontainers/pkg/vcmock/mock_test.go
virtcontainers/pkg/vcmock/sandbox.go
virtcontainers/pkg/vcmock/types.go
virtcontainers/sandbox.go
```

From the files in the PR: https://github.com/bergwolf/kata-runtime/blob/6eb92805dd13ba4bd9e37cbe8a48f1b53c363065/virtcontainers/agent.go#L1-L15

See https://github.com/kata-containers/runtime/issues/227
The correct fix might be to update/fix the license headers in those files...

**Comment by bergwolf** [comment]
@grahamwhaley That's what I meant by "block all PRs until the switch is done", because it is now required to update the license header for any PR that touches the improperly licensed files.

I think the sequence to move forward is to send separate PRs to fix the license headers and get them merged **first**, and then merge this PR to make sure new any added files are properly licensed.

Right now, it seems that we are merging tests first and catching up with implementations.

**Comment by bergwolf** [comment]
And CI for any PR in runtime repo will fail until kata-containers/runtime#227 is fixed.

**Comment by grahamwhaley** [comment]
Well, I understand :-) Yes, we are currently checking any files that are modified as well as newly added.
options:
1) fix the files that are touched in PRs right now whilst we prep the PR for vc SPDX headers
1) temporarily change the check so it only checks newly added files
1) turn the check off completely
1) write and land the vc SPDX update PR asap

I favour maybe a combination of 2 and 4, if we cannot do 4 very quickly... I can jump on (4) in a few hours time....

**Comment by bergwolf** [comment]
@grahamwhaley I'm in favor of any option that can unblock other PRs quickly ;) Thanks!

**Comment by grahamwhaley** [comment]
See https://github.com/kata-containers/runtime/pull/231

**Pull request merged**